p2p/src/sync/peer.rs

```diff
@@ -133,24 +225,15 @@ where
             })
             .await??;
 
-        peer_state.num_blocks_to_send += block_ids.len();
-        self.blocks_queue.extend(block_ids.into_iter().map(|id| (peer, id)));
+        self.blocks_queue.extend(block_ids.into_iter());
```

Comment:
What happens if we receive the same request twice? `self.blocks_queue` is a `VecDeque`.

Reply:
In that case the node sends the requested blocks multiple times, but I'm not sure if a simple check would make a difference here. After all, a malicious peer can request different blocks trying to DoS our node.

Now I'm thinking if we should only allow consecutive block requests. So asking for the blocks with `1, 2, 3, 4, 5` indexes is OK, but `1, 3, 5` isn't. I'm not sure if it is a too severe restriction though.

Reply:
A more reasonable check is whether they already have it. We should know.

Reply:
Also keep in mind that bandwidth dos doesn't make sense, because we don't disallow unknown messages in the interest of upgrades, and it's easy to create a new node and just keep pulling blocks.

Reply:
Fixed.